### PR TITLE
[A0-1605]  Begin OmnitelK Exporter implementation

### PR DESCRIPTION
**Pull request by tigrannajaryan**
This begins implementation of OmnitelK exporter. See design doc:

https://docs.google.com/document/d/1eGa9e08rfqJWvWHRcP-8u4y1dj1FqWpOGQS86Le1Om8/edit?ts=5d72a343#

The exporter is not yet enabled, the code added in this PR is reachable only from tests.

What is done:
- Added omnitelk.proto definition.
- Added "make generate-protobuf" target to generate proto implementation.
- Encoder and ShardEncoder types. Encoder and ShardEncoder
  are covered by tests reasonably well (but need more).
- Renamed .idea files to reflect new repo name.

Todo:
- Implement Exporter, add tests for it.
- Implement a mock backend and add end-to-end tests.
- Implement Client via gRPC.
- Implement Config and Factory for the exporter.

Testing done: make test

### CODE REVIEW

File: exporter/omnitelk/client.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package omnitelk
+
+import omnitelk "github.com/Omnition/omnition-opentelemetry-service/exporter/omnitelk/gen"
+
+type client interface {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/client.go]: Add comment to explain the purpose of the interface.

File: exporter/omnitelk/config.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+// Config contains the main configuration options for the OmnitelK exporter
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"`
+
+	// Endpoint of the server to connect to
+	Endpoint string `mapstructure:"endpoint"`
+
+	// Number of streams to use concurrently. Default value is 1.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/config.go]: What "streams"?

File: exporter/omnitelk/encoder.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+	"github.com/gogo/protobuf/proto"
+	jaeger "github.com/jaegertracing/jaeger/model"
+	"go.opencensus.io/stats/view"
+	"go.uber.org/zap"
+
+	omnitelk "github.com/Omnition/omnition-opentelemetry-service/exporter/omnitelk/gen"
+)
+
+type recordConsumeFunc func(record *omnitelk.EncodedRecord, shard *ShardInMemConfig)
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Why is this declared here while spanProcessFailFunc is declared in shard_encoder.go?

File: exporter/omnitelk/encoder.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	metricViews []*view.View
+
+	// Hooks to publish events to.
+	hooks *kinesisHooks
+
+	logger *zap.Logger
+}
+
+// Options are the options to be used when initializing an Encoder.
+type Options struct {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: These options need doc comments.

File: exporter/omnitelk/encoder.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	Name                  string
+	QueueSize             int
+	NumWorkers            int
+	MaxListSize           int
+	BatchFlushInterval    time.Duration
+	MaxAllowedSizePerSpan int
+}
+
+// NewEncoder returns a Encoder implementation that shards, batches and encodes spans.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]:
```suggestion
// NewEncoder returns an Encoder implementation that shards, batches and encodes spans.
```

File: exporter/omnitelk/encoder.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+// onRecordReady is called when encoded records are ready.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: How about spanProcessFail?

File: exporter/omnitelk/encoder.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+func NewEncoder(
+	o *Options,
+	logger *zap.Logger,
+	onRecordReady recordConsumeFunc,
+	spanProcessFail spanProcessFailFunc,
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]:
```suggestion
	onSpanProcessFail spanProcessFailFunc,
```

File: exporter/omnitelk/encoder.go [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+		onSpanProcessFail: spanProcessFail,
+	}
+
+	return s
+}
+
+// SetConfig sets the sharding configuration. Normally called first during startup
+// and can be called multiple times later when new config is available. Safe to call
+// while processing is in progress.
+func (e *Encoder) SetConfig(config *omnitelk.ShardingConfig) error {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Would be nice to a test that randomly races SetConfig and Encode since that's what will happen in real usage.

File: exporter/omnitelk/encoder.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+	e.stopShardEncoders()
+
+	cfg := NewShardingInMemConfig(config)
+	e.config.Store(cfg)
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: `e.config` seems unused.

File: exporter/omnitelk/encoder.go [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+			hooks,
+			uint64(o.MaxListSize),
+			o.BatchFlushInterval,
+			e.onRecordReady,
+			e.onSpanProcessFail,
+		))
+	}
+
+	e.encoders = encoders
+	e.hooks = newKinesisHooks(o.Name, "")
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Why do we have this? What does empty shard if mean?

File: exporter/omnitelk/encoder.go [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	e.semaphore = nil
+
+	maxReceivers, _ := strconv.Atoi(os.Getenv("MAX_KINESIS_RECEIVERS"))
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Shouldn't this be a config option?

File: exporter/omnitelk/encoder.go [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+	for _, sp := range e.encoders {
+		sp.Start()
+	}
+
+	return nil
+}
+
+// Stop flushes queues and stops exporters.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Unclear comment.

File: exporter/omnitelk/encoder.go [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+		<-e.semaphore
+	}
+}
+
+// EncodeSpan encodes a span and emits resulting batches as encoded records
+// via onRecordReady or onSpanProcessFail callbacks.
+func (e *Encoder) EncodeSpan(span *jaeger.Span) error {
+	e.hooks.OnSpanEnqueued()
+	e.acquire()
+	go e.processSpan(span)
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Consider converting to a size-bound channel instead of a semaphore and goroutines in a future PR.

File: exporter/omnitelk/encoder.go [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+// Stop flushes queues and stops exporters.
+func (e *Encoder) Stop() {
+	e.stopShardEncoders()
+
+	if e.semaphore != nil {
+		close(e.semaphore)
+	}
+}
+
+func (e *Encoder) acquire() {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Perhaps give a better name to acquire/release to make the purpose clear.

File: exporter/omnitelk/encoder.go [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+	return nil
+}
+
+func (e *Encoder) processSpan(span *jaeger.Span) {
+	defer e.release()
+
+	// Update stats.
+	e.hooks.OnSpanDequeued()
+
+	// Find the right shard encoder base don trace id.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]:
```suggestion
	// Find the right shard encoder based on trace id.
```

File: exporter/omnitelk/encoder.go [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+		e.logger.Error("failed to get producer/shard", zap.Error(err),
+			zap.String("traceID", span.TraceID.String()))
+		e.onSpanProcessFail([]*jaeger.Span{span}, FailedNotRetryable)
+		return
+	}
+
+	// Marshal to calculate the size.
+	encoded, err := proto.Marshal(span)
+	if err != nil {
+		e.logger.Error("failed to marshal: ", zap.Error(err))
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Can this cause a log storm if we receive a lot of bad data? Do we need it if we already report it via onSpanProcessFail?

File: exporter/omnitelk/encoder.go [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+		size = len(encoded)
+	}
+	// TODO: See if we can encode only once and Encode encoded span on the shard producer.
+	// shard producer will have to arrange the bytes exactly as protobuf marshaller would
+	// encode a SpanList object.
+	se.Encode(span, uint64(size))
+}
+
+func (e *Encoder) getShardEncoder(partitionKey string) (*ShardEncoder, error) {
+	for _, sp := range e.encoders {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder.go]: Consider using binary search in a future PR. Each belongsToShard may be relatively slow (does md5, than big.Int comparisons) so binary search may be warranted.

File: exporter/omnitelk/encoder_test.go [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+	omnitelk "github.com/Omnition/omnition-opentelemetry-service/exporter/omnitelk/gen"
+)
+
+var testConfig = &omnitelk.ShardingConfig{
+	ShardDefinitions: []*omnitelk.ShardDefinition{
+		{ShardId: "1", StartingHashKey: minHashKey.Bytes(), EndingHashKey: maxHashKey.Bytes()},
+	},
+}
+
+type encoderSink struct {
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder_test.go]: Add a doc comment.

File: exporter/omnitelk/encoder_test.go [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+		// Use large flash interval to avoid flushing.
+		e := createEncoder(t, time.Second*10, sink)
+
+		// Encode a batch of spans.
+		const spanCount = 100
+		for j := 0; j < spanCount; j++ {
+			e.EncodeSpan(createSpan(j))
+		}
+
+		// Stop without failing for encoding.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/encoder_test.go]: Unclear comment.

File: exporter/omnitelk/exporter.go [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	var exportErr error
+	for _, span := range pBatch.GetSpans() {
+		if span.Process == nil {
+			span.Process = pBatch.Process
+			// Encode the span. Wait if encoder is busy. This will apply backpressure.
+			exportErr = e.encoder.EncodeSpan(span)
+			if exportErr != nil {
+				// TODO: count failed encodings.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/exporter.go]:
```suggestion
				// TODO: count failed encode operations.
```

File: exporter/omnitelk/exporter.go [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+func (e *Exporter) processRetryQueue() {
+	// Read spans from the retry queue and try to encode them again.
+
+	// Here we don't need to watch "done" signal since Stop() will close spansToRetry
+	// channel when stopping.
+	for spans := range e.spansToRetry {
+		for _, span := range spans {
+			if err := e.encoder.EncodeSpan(span); err != nil {
+				// TODO: count failed encodings.
```
Comment (tigrannajaryan) [inline exporter/omnitelk/exporter.go]: ditto

File: exporter/omnitelk/exporter.go [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+			}
+		}
+	}
+}
+
+// onRecordReady is called when an encoded record is ready. This function will send it
+// to the server.
+func (e *Exporter) onRecordReady(record *omnitelk.EncodedRecord, shard *ShardInMemConfig) {
+	e.client.Send(record, shard)
```
Comment (tigrannajaryan) [inline exporter/omnitelk/exporter.go]: Add a TODO to make sure this is handled properly when return value is added to `Send`.

File: exporter/omnitelk/shard_encoder.go [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+	"bytes"
+	"compress/gzip"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	jaeger "github.com/jaegertracing/jaeger/model"
+	model "github.com/omnition/opencensus-go-exporter-kinesis/models/gen"
```
Comment (tigrannajaryan) [inline exporter/omnitelk/shard_encoder.go]:
```suggestion
	kinesismodel "github.com/omnition/opencensus-go-exporter-kinesis/models/gen"
```

File: exporter/omnitelk/shard_encoder.go [hunk 24]
```diff
@@ -0,0 +1,263 @@
...
+	"go.uber.org/zap"
+
+	omnitelk "github.com/Omnition/omnition-opentelemetry-service/exporter/omnitelk/gen"
+)
+
+const avgBatchSize = 1000
```
Comment (tigrannajaryan) [inline exporter/omnitelk/shard_encoder.go]: What's the unit of size?

File: exporter/omnitelk/encoder.go (hunk 10 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/encoder.go]: These are hooks that are not shard dependent.

File: exporter/omnitelk/encoder.go [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+	}
+}
+
+func (e *Encoder) acquireWorker() {
+	// Acquires a worker. Concurrency of workers may be limited via semaphore
+	// if defined so when Encoder is created. In that case this function will block
+	// until the number of currently running workers goes below concurrency limit
+	// defined by NumWorkers option.
+	if e.semaphore != nil {
+		e.semaphore <- struct{}{}
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/encoder.go]: All re-worked and fixed bugs.

File: exporter/omnitelk/exporter.go [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+
+// Stop exporter. Exporter cannot be started after it is stopped. Safe to call
+// concurrently with other public functions.
+func (e *Exporter) Stop() error {
+	// Close channels to stop processing.
+	close(e.spansToRetry)
+	close(e.done)
+
+	// Stop Encoder.
+	e.encoder.Stop()
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/exporter.go]: Added graceful stopping.

File: exporter/omnitelk/omnitelk.proto [hunk 27]
```diff
@@ -0,0 +1,92 @@
+syntax = "proto3";
+
+package omnitelk;
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/omnitelk.proto]: Done.

**Comment by tigrannajaryan** [comment]
Please have another look. I removed the Exporter, Client, Config and Factory from the PR to make it smaller. Added a bit more tests, docs and diagrams.
This needs more tests to ensure it is rock solid. I will add in a future PR to avoid making this even bigger.

File: exporter/omnitelk/shard_encoder.go [hunk 28]
```diff
@@ -0,0 +1,358 @@
...
+	se.spans.Spans = se.spans.Spans[:0]
+	se.accumulatedSize = 0
+}
+
+func (se *ShardEncoder) isStopped() bool {
+	return atomic.LoadUint32(&se.stopped) > 0
+}
+
+// compress is unsafe for concurrent usage. caller must protect calls with mutexes.
+func (se *ShardEncoder) compress(in []byte) ([]byte, error) {
```
Comment (owais) [inline exporter/omnitelk/shard_encoder.go]: I originally wrote it like this but perhaps we should change it to so that this function accepts `out []byte` as a second argument. Returning a new buffer every from the function will probably result in it escaping to heap. If we pass the `out` buffer to this func from outside, we can then try to reuse the same buffer like the shared `spans` slice without allocating in every compress() call or could use something like sync.Pool to leverage object pooling. 

Just putting it out there for now. We don't necessarily need to do it right way.

File: exporter/omnitelk/exporter.go [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+		done:         make(chan interface{}),
+		client:       client,
+	}
+	e.encoder = NewEncoder(&Options{}, logger, e.onRecordReady, e.onSpanProcessFail)
+	return e
+}
+
+// Start exporter.
+func (e *Exporter) Start() {
+	go e.connectAndGetConfig()
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/exporter.go]: I removed the Exporter from this PR to make it smaller, but the answer is yes, it will. We don't have much choice, that's how otelsvc is designed. As @pjanotti correctly pointed out a better approach is to add ability to make Health Check probe depend on the state of other components and only declare the service ready after Exporter successfully connects.
Absent that we have to accept and accumulate the spans until connection is established. We will not accumulate much, we have a small input queue. If connection is not established quickly we will start blocking ConsumeTraceData which will exhort backpressure or eventually hit memory limiter.

File: exporter/omnitelk/shard_encoder.go [hunk 30]
```diff
@@ -0,0 +1,358 @@
...
+	return uint(size)
+}
+
+// Flush accumulated spans. Encodes and emits via onSpanProcessFail on success or
+// via onSpanProcessFail on error.
+func (se *ShardEncoder) Flush() {
+	// Lock the encode. Note that we Unlock before returning or before calling
+	// onSpanProcessFail or onRecordReady callbacks to hold the lock for as short
+	// as possible.
+	se.Lock()
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/shard_encoder.go]: Unfortunately no, we have multiple exit paths form the function and Unlock() needs to happen at the right time.

File: exporter/omnitelk/shard_encoder.go [hunk 31]
```diff
@@ -0,0 +1,358 @@
...
+			se.onSpanProcessFail([]*jaeger.Span{span}, FailedNotRetryable)
+			return 0, errors.New("spans is too big")
+		}
+	}
+	return size, nil
+}
+
+func (se *ShardEncoder) accumulateEncodedSpan(span *jaeger.Span, size int) {
+	// Acquire lock to shared data.
+	se.Lock()
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/shard_encoder.go]: Nope :-)
Need to Unlock before onSpanProcessFail and before calling flush so it has to be manual. We cannot keep the lock until return.

File: exporter/omnitelk/shard_encoder.go (hunk 28 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/shard_encoder.go]: Good idea, I can explore it in the future.

File: exporter/omnitelk/hooks.go [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+type tagCache struct {
+	sync.RWMutex
+	tags map[tag.Key]map[string]tag.Mutator
+}
+
+type kinesisHooks struct {
```
Comment (pjanotti) [inline exporter/omnitelk/hooks.go]:
```suggestion
type observabilityHooks struct {
```

or something else but there is nothing related to kinesis here.

File: exporter/omnitelk/shard_encoder.go [hunk 33]
```diff
@@ -0,0 +1,358 @@
...
+	// Gzip writer to compressed encoded spans.
+	gzipWriter *gzip.Writer
+
+	// Mutex to protect spans, accumulatedSize and gzipWriter fields which are used
+	// concurrently.
+	sync.RWMutex
+
+	// Done signal, used to interrupt internal goroutine.
+	done chan interface{}
+	// Stopped flag, is set to non-zero simultaneously with done signal.
```
Comment (pjanotti) [inline exporter/omnitelk/shard_encoder.go]: nit: add blank line

File: exporter/omnitelk/shard_encoder.go [hunk 34]
```diff
@@ -0,0 +1,358 @@
...
+	}
+
+	if err := se.gzipWriter.Close(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (se *ShardEncoder) flushPeriodically() {
+	// TODO: Start timer after last flush instead of periodic.
```
Comment (pjanotti) [inline exporter/omnitelk/shard_encoder.go]: Isn't that the case of simply calling https://godoc.org/time#Timer.Reset after the flush completes?

File: exporter/omnitelk/shard_encoder.go [hunk 35]
```diff
@@ -0,0 +1,358 @@
...
+	encoded, err := proto.Marshal(span)
+	if err != nil {
+		se.logger.Error("failed to marshal", zap.Error(err))
+		se.onSpanProcessFail([]*jaeger.Span{span}, FailedNotRetryable)
+		return 0, err
+	}
+	size = len(encoded)
+	if uint(size) > se.maxAllowedSizePerSpan {
+		// Span is too big.
+		se.hooks.OnXLSpanDropped(size)
```
Comment (pjanotti) [inline exporter/omnitelk/shard_encoder.go]: Perhaps separate XLSpanTruncated vs actually dropped below.

File: exporter/omnitelk/encoder.go [hunk 36]
```diff
@@ -0,0 +1,459 @@
...
+
+	// Non-zero value indicates that Encoder is stopping.
+	isStoppingFlag int32
+
+	// Channel to signal to workers to stop immediately.
+	workerHardStopSignal chan interface{}
+
+	// Points to immutable shardEncoders. Unfortunately atomic.Value is not strongly
+	// typed, so be careful to always interpret the value as shardEncoders.
+	encoders atomic.Value
```
Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: nit: I find confuse the name encoders here, since this is Encoder type. Perhaps `sharders` to avoid conflict with the `shardEncoders` type?

File: exporter/omnitelk/encoder.go [hunk 37]
```diff
@@ -0,0 +1,459 @@
...
+//            +---------+  --/  +------+          +--------+
+// EncodeSpans|Span     |-/               ----->
+// ---------->|Batches  |         ...                ...
+//            |Queue    |-\               ----->
+//            +---------+  --\  +------+          +--------+
+//                            ->|Worker|  ----->  |Shard   |--->onRecordReady
+//                              +------+          |Encoder |--->onSpanProcessFail
+//                                                +--------+
+//
+type Encoder struct {
```
Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: Does `Encoder` needs to be public?

File: exporter/omnitelk/encoder.go [hunk 38]
```diff
@@ -0,0 +1,459 @@
...
+	// Wait until all pending spans are processed but no more than maxDrainDuration.
+	// spansToProcessWg counts the number of span batches that are pending processing.
+	if !waitGroupTimeout(&e.spansToProcessWg, maxDrainDuration) {
+		e.logger.Warn("timed out waiting for Encoder to drain spans")
+	}
+
+	// We either successfully drained spans from our queue to shard encoders or timed out.
+	// As a result we may have some spans accumulated in shard encoders. Now flush them.
+	encoders := e.encoders.Load().(shardEncoders)
+	for _, sp := range encoders {
```
Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: I would recommend using a WaitGroup here and call each `Flush` in a goroutine: the rational is that this is time sentitive and there can be network IO on the flush so making then concurrently makes sense.

Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: nit: `sp` doesn't match anything here

File: exporter/omnitelk/encoder.go [hunk 39]
```diff
@@ -0,0 +1,459 @@
...
+// Stops shard encoders. Signals to encoders to stop but does not wait for encoders to
+// actually stop, so they may continue running briefly while they get the signal.
+// Any currently processing spans that are not yet emitted via onRecordReady will be
+// returned by calling onSpanProcessFail func.
+func (e *Encoder) stopShardEncoders() {
+	encoders := e.encoders.Load().(shardEncoders)
+	e.encoders.Store(make(shardEncoders, 0))
+
+	// Tell all encoders to stop.
+	for _, sp := range encoders {
```
Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: nit: `sp` as mentioned before doesn't match anything around here, will ignore others

File: exporter/omnitelk/telemetry.go [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+	"go.opencensus.io/tag"
+)
+
+// Keys and stats for telemetry.
+var (
+	tagExporterName, _ = tag.NewKey("exporter")
+	tagShardID, _      = tag.NewKey("shard_id")
+	tagFlushReason, _  = tag.NewKey("flush_reason")
+
+	statXLSpansBytes = stats.Int64("kinesis_xl_span_size", "size of spans bigger than max support size", stats.UnitBytes)
```
Comment (pjanotti) [inline exporter/omnitelk/telemetry.go]: nit: these are actually not related to kinesis

File: exporter/omnitelk/shard_encoder.go (hunk 34 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/shard_encoder.go]: Yes, probably. We don't have the timer pointer in Flush() func, so it will require a bit of refactoring. I am leaving this for later.

File: exporter/omnitelk/shard_encoder.go (hunk 35 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/shard_encoder.go]: Renaming this to XLSpanTruncated. Will add dropped in a separate PR about telemetry since there are more missing bits.

File: exporter/omnitelk/encoder.go (hunk 37 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/encoder.go]: No. Fixing.

File: exporter/omnitelk/encoder.go (hunk 38 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/encoder.go]: Good point. Adding.

File: exporter/omnitelk/telemetry.go (hunk 40 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/telemetry.go]: Fixing.
